toolbox/hgfs/archive.go

```diff
@@ -229,42 +229,37 @@ func archiveRead(u *url.URL, tr *tar.Reader) error {
 
 // archiveWrite writes the contents of the given source directory to the given tar.Writer.
 func archiveWrite(u *url.URL, tw *tar.Writer) error {
-	dir := filepath.Dir(u.Path)
 
 	return filepath.Walk(u.Path, func(file string, fi os.FileInfo, err error) error {
 		if err != nil {
-			return filepath.SkipDir
+			return err
 		}
 
-		name := strings.TrimPrefix(file, dir)[1:]
-
-		header, _ := tar.FileInfoHeader(fi, name)
-
-		header.Name = name
-
-		if header.Typeflag == tar.TypeDir {
-			header.Name += "/"
+		header, err := tar.FileInfoHeader(fi, fi.Name())
+		if err != nil {
+			return err
 		}
+		header.Name = strings.TrimPrefix(
+			strings.Replace(file, filepath.Dir(u.Path), "", -1),
+			string(filepath.Separator))
 
-		var f *os.File
-
-		if header.Typeflag == tar.TypeReg && fi.Size() != 0 {
-			f, err = os.Open(file)
-			if err != nil {
-				if os.IsPermission(err) {
-					return nil
-				}
-				return err
-			}
+		if header.Typeflag != tar.TypeReg {
+			return nil
 		}
 
-		_ = tw.WriteHeader(header)
+		if err := tw.WriteHeader(header); err != nil {
+			return err
```

Comment:
Was there a case where WriteHeader returned an error related to the header itself?  Otherwise, if the pipe was closed, then the io.Copy below would return the same error.